app/services/action_runner.py

```diff
@@ -18,20 +21,40 @@
 logger = logging.getLogger(__name__)
 
 
-async def execute_action(integration_id: str, action_id: str, config_overrides: dict = None):
+async def execute_action(integration_id: str, action_id: str, config_overrides: dict = None, config_data: dict = None):
     """
     Interface for executing actions.
     :param integration_id: The UUID of the integration
     :param action_id: "test_auth", "pull_observations", "pull_events"
     :param config_overrides: Optional dictionary with configuration overrides
+    :param config_data: Optional dictionary with an already-set configuration (to be passed to action handler as it is)
     :return: action result if any, or raise an exception
     """
     logger.info(f"Executing action '{action_id}' for integration '{integration_id}'...")
-    try:  # Get the integration config from the portal
-        async for attempt in stamina.retry_context(on=httpx.HTTPError, wait_initial=1.0, wait_jitter=5.0, wait_max=32.0):
-            with attempt:
-                # ToDo: Store configs and update it on changes (event-driven architecture)
-                integration = await _portal.get_integration_details(integration_id=integration_id)
+    try:  # Get the integration config from cache
+        state_manager = IntegrationStateManager()
```

Comment:
I agree about storing configurations in Redis and avoiding extra calls to the portal, but I was thinking that the integration `state` and the `configuration` are, conceptually, different things. So maybe we can have a separate `IntegrationConfigurationsManager` to retrieve configurations, and maybe the logic to either get those from Redis or the portal could be encapsulated there.